src/sentry/issues/escalating.py

```diff
@@ -29,16 +39,58 @@
 
 
 def query_groups_past_counts(groups: List[Group]) -> List[GroupsCountResponse]:
-    """Query Snuba for the counts for every group bucketed into hours"""
-    offset = 0
+    """Query Snuba for the counts for every group bucketed into hours.
+
+    It optimizes the query by guaranteeing that we look at group_ids that are from the same project id.
+    This is important for Snuba as the data is stored in blocks related to the project id.
+
+    We maximize the number of projects and groups to reduce the total number of Snuba queries.
+    Each project may not have enough groups in order to reach the max number of returned
+    elements (QUERY_LIMIT), thus, projects with few groups should be grouped together until
+    we get at least a certain number of groups.
+
+    NOTE: Groups with less than the maximum number of buckets (think of groups with just 1 event or less
+    than 7 days old) will skew the optimization since we may only get one page and less elements than the max
+    QUERY_LIMIT.
+    """
+
     all_results = []
     start_date, end_date = _start_and_end_dates()
-    project_ids, group_ids = _extract_project_and_group_ids(groups)
+    group_ids_by_project = _extract_project_and_group_ids(groups)
+    proj_ids, group_ids = [], []
+    processed_projects = 1
+    total_projects_count = len(group_ids_by_project)
+
+    for proj_id in group_ids_by_project.keys():
+        _group_ids: List[int] = group_ids_by_project[proj_id]
+        # Add them to the list of projects and groups to query
+        proj_ids.append(proj_id)
+        group_ids += _group_ids
+        processed_projects += 1
+        # We still have room for more projects and groups
+        if (
+            processed_projects < total_projects_count
+            and len(_group_ids) < QUERY_LIMIT / BUCKETS_PER_GROUP
+        ):
+            continue
+
+        all_results += _query_with_pagination(proj_ids, group_ids, start_date, end_date)
+        # We're ready for a new set of projects and ids
+        proj_ids, group_ids = [], []
 
+    return all_results
+
+
+def _query_with_pagination(
+    project_ids: List[int], group_ids: List[int], start_date: datetime, end_date: datetime
+) -> List[GroupsCountResponse]:
+
+    all_results = []
+    offset = 0
     while True:
```

Comment:
Is this going to issue a number of queries that we can determine aren't necessary by logic? e.g. if LIMIT is 10k and you receive a result with 5k, we should break from the loop rather than query again and get 0 results right?

Reply:
Great suggestion! I've addressed it.